stock_shift_residual_qty/models/stock_picking.py

```diff
@@ -0,0 +1,89 @@
+# Copyright 2020 David BEAL @ Akretion
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import _, models
+from odoo.exceptions import UserError
+
+
+class StockPicking(models.Model):
+    _inherit = "stock.picking"
+
+    def button_validate(self):
+        res = super().button_validate()
+        wh_ids = [x.picking_type_id.warehouse_id.id for x in self]
+        internal_types = self.env["stock.picking.type"].search(
+            [("warehouse_id", "in", wh_ids), ("code", "=", "internal")]
+        )
+        pick_types = {x.warehouse_id: x.id for x in internal_types}
+        if not pick_types:
```

Comment:
Maybe you should test if there any "self._get_residual_product_and_qty()" before doing theses checks ?

Reply:
In case of many pickings I prefer make just one search in db before `rec in self:` to limit the number of queries

Reply:
Finally hardly refactor, this confused part is not kept